### PR TITLE
JIRA TRAFODION-1798 (ddl xns) and few other fixes, details below.

### PR DESCRIPTION
**Pull request by anoopsharma00**
-- support for sql part of ddl xns. Section 1 of JIRA TRAFODION-1798
-- cqd ddl_transactions to enable or disable ddl xns.
   Default is currently off. Once it is tested, it will be turned on.
   Dev regressions are run with cqd set to ON
-- get stmts run with read committed to get changes in current xns
-- support for where preds with get stmts
-- scan to pass in transid even if running with read uncommitted access.
   This enables rows modified in current xn to be returned.
-- cleanup no longer return multiple duplicate error messages if
   objects id is not found.
-- cleanup no longer includes internallay created schemas (_HV_ , _HB_)
   during cleanup operations.
-- Correct error msg was not getting returned if an invalid index
   existed in table and the same index was created again.
-- init traf, drop md views was giving an error if views didnt exist.
   That has been fixed.
-- regressions with -diff option now show original file timestamps
   instead of the timestamp when the diff command was run.

### CODE REVIEW

**Comment by shangshengtung** [comment]
The changes on TM side look good.

**Comment by prashanth-vasudev** [comment]
Code changes look good to me. Thank You.

File: core/sqf/src/seatrans/hbase-trx/src/main/java/org/apache/hadoop/hbase/client/transactional/TransactionalScanner.java [hunk 1]
```diff
@@ -171,19 +171,21 @@ protected boolean nextScanner(final boolean done) throws IOException{
...
-        this.currentRegion = ttable.getRegionLocation(this.currentBeginKey).getRegionInfo();
+        this.currentRegion = ttable.getRegionLocation(this.currentBeginKey, false).getRegionInfo();
         this.currentEndKey = this.currentRegion.getEndKey();
 
-        if(LOG.isTraceEnabled()) LOG.trace("Region Info: " + currentRegion.getRegionNameAsString());
-        if(this.currentEndKey != HConstants.EMPTY_END_ROW)
-            this.currentEndKey = TransactionManager.binaryIncrementPos(currentRegion.getEndKey(), -1);
+        if(LOG.isTraceEnabled()) LOG.trace("nextScanner() txID: " + ts.getTransactionId() + " Region Info: " + currentRegion.getRegionNameAsString()+ 
+                         "currentBeginKey: " + Hex.encodeHexString(this.currentBeginKey) + "currentEndKey: " + Hex.encodeHexString(this.currentEndKey));
+        //if(this.currentEndKey != HConstants.EMPTY_END_ROW)
```
Comment (selvaganesang) [inline core/sqf/src/seatrans/hbase-trx/src/main/java/org/apache/hadoop/hbase/client/transactional/TransactionalScanner.java]: Consider removing the commented out lines

File: core/sql/executor/ex_transaction.cpp [hunk 2]
```diff
@@ -1316,11 +1309,29 @@ short ExTransTcb::work()
...
+            if (cmpStatus == ExSqlComp::ERROR)
+              {
+                handleErrors(pentry_down, NULL, 
+                             (ExeErrorCode)(-EXE_CANT_COMMIT_OR_ROLLBACK));
+ 
+                return -1;
+              }
+          }
+        
+        cliGlobals->currContext()->ddlStmtsExecuted() = FALSE;
```
Comment (selvaganesang) [inline core/sql/executor/ex_transaction.cpp]: The ddlStmtsExecuted flag will be set to TRUE if there is any error on commit or rollback. How can the user process recover from this,

File: core/sql/cli/Context.h [hunk 3]
```diff
@@ -472,6 +477,10 @@ class ContextCli : public ExGod {
   bool cbServerInUse_;
   NABoolean dropInProgress_;
 
+  // set to true if ddl stmts are issued.
+  // Reset at begin and commit Xn. Used to do NATable invalidation
+  // for ddl stmts issued within a transaction.
```
Comment (DaveBirdsall) [inline core/sql/cli/Context.h]: Glad you are doing this. This has often been a sore point.

File: core/sql/executor/ExExeUtilGet.cpp [hunk 4]
```diff
@@ -3173,6 +3149,21 @@ short ExExeUtilGetHbaseObjectsTcb::work()
           }
           break;
 
+        case EVAL_EXPR_:
```
Comment (DaveBirdsall) [inline core/sql/executor/ExExeUtilGet.cpp]: This is a very nice feature. Thanks for adding this. I've always wondered though why GET statements aren't simply SELECTs. Would be nice to embed them within other SELECTs for example. There have been times when I've wanted to join them to other things too.

File: core/sql/optimizer/NARoutineDB.h [hunk 5]
```diff
@@ -50,7 +50,7 @@ class NARoutineDB : public NAKeyLookup<NARoutineDBKey,NARoutine>
 
 public:
   // This same enum exists in NATableDB, should probably only have one copy
-  enum QiScope { REMOVE_FROM_ALL_USERS = 100, REMOVE_MINE_ONLY };
+  //  enum QiScope { REMOVE_FROM_ALL_USERS = 100, REMOVE_MINE_ONLY };
```
Comment (DaveBirdsall) [inline core/sql/optimizer/NARoutineDB.h]: Could get rid of these comments?

File: core/sql/parser/StmtDDLAlter.cpp [hunk 6]
```diff
@@ -1971,6 +1971,17 @@ StmtDDLAlterTableAddColumn::synthesize()
...
+    ElemDDLColDefArray ColDefArray = getColDefArray();
+    ElemDDLColDef *pColDef = ColDefArray[0];
+    if (NOT pColDef->getColumnFamily().isNull())
+      {
+        //TEMPTEMP 
+        // Currently, DTM doesnt handle add columns with an explicit 
+        // column family as a transactional operation.
+        // Do not use ddl xns until that bug is fixed.
+        setDdlXns(FALSE);
+      }
```
Comment (DaveBirdsall) [inline core/sql/parser/StmtDDLAlter.cpp]: How about ALTER TABLE/INDEX ALTER HBASE_OPTIONS?

File: core/sql/sqlcomp/CmpSeabaseDDLcleanup.cpp [hunk 7]
```diff
@@ -1089,7 +1092,7 @@ short CmpSeabaseMDcleanup::cleanupOrphanObjectsEntries(ExeCliInterface *cliInter
   NABoolean errorSeen = FALSE;
 
   // find out all entries which do not have corresponsing hbase objects
-  str_sprintf(query, "select object_uid, trim(catalog_name) || '.'  || trim(schema_name) || '.' || trim(object_name)  from %s.\"%s\".%s where catalog_name = '%s' and schema_name not in ( '_MD_', '_REPOS_', 'PRIVMGR_MD') and (object_type = 'BT' or object_type = 'IX') ",
+  str_sprintf(query, "select object_uid, trim(catalog_name) || '.'  || trim(schema_name) || '.' || trim(object_name)  from %s.\"%s\".%s where catalog_name = '%s' and schema_name not in ( '_MD_', '_REPOS_', '_PRIVMGR_MD_') and schema_name not like '|_HV|_%%' escape '|'  and schema_name not like '|_HB|_%%' escape '|' and (object_type = 'BT' or object_type = 'IX') ",
```
Comment (DaveBirdsall) [inline core/sql/sqlcomp/CmpSeabaseDDLcleanup.cpp]: Are schema names beginning with _HV and _HB reserved? If not, then this statement might mistakenly filter out user schemas (like "_HVAC").

File: core/sql/sqlcomp/CmpSeabaseDDLcommon.cpp [hunk 8]
```diff
@@ -8167,35 +8239,71 @@ short CmpSeabaseDDL::executeSeabaseDDL(DDLExpr * ddlExpr, ExprNode * ddlNode,
...
+      ((ddlNode->getOperatorType() == DDL_DROP_SCHEMA) ||
+       (ddlNode->getOperatorType() == DDL_CLEANUP_OBJECTS) ||
+       (ddlNode->getOperatorType() == DDL_ALTER_TABLE_ADD_CONSTRAINT_PRIMARY_KEY) ||
+       (ddlNode->getOperatorType() == DDL_ALTER_TABLE_ALTER_COLUMN_SET_SG_OPTION) ||
+       (ddlNode->getOperatorType() == DDL_CREATE_INDEX) ||
+       (ddlNode->getOperatorType() == DDL_POPULATE_INDEX) ||
+       (ddlNode->getOperatorType() == DDL_CREATE_TABLE) ||
+       (ddlNode->getOperatorType() == DDL_ALTER_TABLE_DROP_COLUMN) ||
+       (ddlNode->getOperatorType() == DDL_DROP_TABLE)))
+    {
```
Comment (DaveBirdsall) [inline core/sql/sqlcomp/CmpSeabaseDDLcommon.cpp]: Earlier I had asked about ALTER TABLE/INDEX ALTER HBASE_OPTIONS... maybe that one should be here too.

**Comment by DaveBirdsall** [comment]
Looks good to me. Some questions though might (or might not!) prompt follow-on work.

**Comment by DaveBirdsall** [comment]
Is this change ready to commit?

**Comment by anoopsharma00** [comment]
It is ready to commit from code changes pov. Code has been merged with latest branch and contains the needed functionality.

**Pull request merged**